src/main/java/org/opensearch/neuralsearch/executors/HybridQueryExecutorCollector.java

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.opensearch.neuralsearch.executors;
+
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+import lombok.Synchronized;
+
+import java.util.Optional;
+import java.util.function.Function;
+
+/**
+ * {@link HybridQueryExecutorCollector} is a generic Collector used by Hybrid Search Query during
+ * Query phase to parallelize sub query's action to improve latency
+ */
+@RequiredArgsConstructor(staticName = "newCollector")
+public final class HybridQueryExecutorCollector<I, R> {
+
+    // will be used as input for all instances of collector generated by newCollector method,
+    // if it is required for collect operation
+    private final I param;
+
+    // getResult should only be called after collector's collect method is invoked.
+    @Getter(onMethod_ = { @Synchronized })
+    private Optional<R> result = Optional.empty();
```

Comment:
we're saving some state in collector. Is it guaranteed that same collector instance is not used second time before this saved state (result) is read by the caller?
The way it's done in core for QueryContext and Collectors - they put Collector Manager into same package with Collector classes, everything in Collector is package private, only CollectorManager methods are public. This way they shielded from scenario when someone reuse same instance of Collector and state is corrupted. 
This also opens up possibility for us to reuse Collector instances in future, if some of them can be redesigned to be stateless.

Reply:
Yes @martin-gaievski Every-time manager will create new instance when newCollector method is called. Let me move collector manager and collector in same package and make it package private if this helps.

Reply:
I think separate package for executors is good, I think even before your latest commit we need to cleanup the `query` package. Can you please do the other way around - restore that `executors` package and keep (or move) executor specific classes in that package. Then we can have `newCollector` method of Manager class as public. It's ok to have Collector.getResult public, Optional is immutable so client should not be able to change it.

Reply:
Ack.